Add timeseries to datapackage

FelixMau opened the pull request.

FelixMau committed:
Clean todos and change collection

FelixMau committed:
Clean todos and change collection

FelixMau committed:
Clean todos and change collection

FelixMau committed:
Only returning found values

Commit:
Add extra paramater column  to facade defaults

Parameter `region` and `year` are necessary to
identify the timeseries profile which is then,
written into column `profile`.

Commit:
Add extra paramater column  to facade defaults

Parameter `region` and `year` are necessary to
identify the timeseries profile which is then,
written into column `profile`.

Commit:
Make code more lean

Its not really necessary to store the results in the object itself
and return it. Also the results can then be added to the parametrized
elements in a simpler way.

Commit:
Make code more lean

Its not really necessary to store the results in the object itself
and return it. Also the results can then be added to the parametrized
elements in a simpler way.

Commit:
Add draft of time series refactoring

This is a first draft for the timeseries refactoring.
Needs to be put into a separate function.

Commit:
First draft for the timeseries refactoring

This needs to be put into sepearate function. Matching between
scalar rows and timeseries column is still missing. Can be matched
via `region` and `year`.

Commit:
First draft for the timeseries refactoring

This needs to be put into sepearate function. Matching between
scalar rows and timeseries column is still missing. Can be matched
via `region` and `year`.

Commit:
Change to hack-a-than exmpale

Commit:
Add comment

FelixMau committed:
refactoring and writing to function

FelixMau committed:
refactoring and writing to function

FelixMau committed:
Moveing refactor_timeseries out of class

The functionality might also be used somewhere else

FelixMau committed:
Adding Profiles to scalars

FelixMau committed:
Adding Profiles to scalars

FelixMau committed:
Merge remote-tracking branch 'origin/add-timeseries-to-datapackage' into add-timeseries-to-datapackage

# Conflicts:
#	data_adapter_oemof/build_datapackage.py

FelixMau committed:
Adding Profiles collumn name to scalars

Moving refactor function outside of class (might be usefull from outside as well)

Rebasing

FelixMau committed:
Adding region to scalars
Rewriting loop to be independet of order and adding region
Adding Timeseries naming for collumns

FelixMau committed:
Adapting building logic and order of parametrized_sequences and adding `profile` column and entries for parametrized_scalars

- Refactor the time series using the `refactor_timeseries` function.
- The `facade adapters` that include a time series will have an additional variable called `profiles`, which is a list of column names for profiles.
- If there is more than one time series or entries in profiles, a warning will be raised, and no time series will be processed (this feature is not yet implemented).
- When creating the scalar data frames, add another column whose name comes from the adapter's profiles variable.
- The entries of this profile column will be created using  `original name of the time series column` + `_` + `region` from the scalar.

FelixMau committed:
Applying changes to get_process call to follow depreciation warning

FelixMau committed:
Applying changes to get_process call to follow depreciation warning

FelixMau committed:
Implementing data_adapter.preprocessing.Adapter class into building datapackage.

This will help improve acessing useful functions that might be implemented later as the Adapter is containing structure as well as processes as well as functionality to modify or grap those

FelixMau committed:
Implementing data_adapter.preprocessing.Adapter class into building datapackage.

This will help improve acessing useful functions that might be implemented later as the Adapter is containing structure as well as processes as well as functionality to modify or grap those

FelixMau committed:
Bugfix

FelixMau committed:
Splitting parametrized sequences into years

FelixMau commented:
I think it will be easier to review if creation of `foreign keys` and saving of the `DataPackage` is moved to another branch to keep PRs small.

FelixMau committed:
Cleaning

henhuy commented on data_adapter_oemof/build_datapackage.py:
```diff
@@ -1,92 +1,185 @@
+        for profile_name in ts_columns:
+            profile_column = df[profile_name].dropna()
+            profile_column = profile_column.explode().to_frame()
+            profile_column["timeindex"] = profile_column.groupby(level=0).cumcount()
+            profile_column = profile_column.reset_index().pivot(
+                index="timeindex", columns="index", values=profile_name
+            )
```
What's happening here?
Hard to understand

henhuy replied:
It feels like this could be simplified.

iuljan replied:
I agree, it's complicated but we didn't find a way to unnest and reformat multiple arrays at once into columns while keeping timeindex order. I added further comments to explain the process.

henhuy commented on data_adapter_oemof/adapters.py:
```diff
@@ -27,13 +27,16 @@ def as_dict(self):
             ),
+            "region": mapper.get("region"),
+            "year": mapper.get("year"),
+            "profile": mapper.get("profile"),
```
not every component has a profile - should only be added to individual adapters

henhuy commented on data_adapter_oemof/build_datapackage.py:
```diff
@@ -1,92 +1,185 @@
-                    ],
+
+            scalars = pd.DataFrame(scalars.to_list())
+            if "profiles" in facade_adapter.__dict__:
```
I think refactoring timeseries can be more generalized and instead mapping of profiles to TS columns should be added - will tell you my idea in a meeting.
Then this hardcoded line could be removed and limiting TS to single column could be removed as well

iuljan replied:
This needs further elaboration

henhuy commented on data_adapter_oemof/build_datapackage.py:
```diff
@@ -1,92 +1,185 @@
+                                process.timeseries.columns.difference(core.TIMESERIES_COLUMNS.keys())[0]
+                                + "_"
+                                + scalars["region"]
+                        )
```
My idea is do move this part to parametrize_dataclass and its mappings (therefore we have to introduce component specific mappings next to default mappings ad change behaviour of mapping a bit).
So that for example for a load component, there would exist a mapping something like this:
```
Default:
  amount: amount
LoadAdapter:
  profile: load
```
The mapper would see that there is no column `load` in scalars and would check if such column exists in timeseries, if so it would simply return the value "load" as this has to be set as profile.
Additionally, we should add a calculation to address region-based profiles (as you already enabled TS like "load-TH" - very good by the way!!!)

henhuy reviewed (changes requested):
My feedback needs further information - lets discuss this in a meeting

henhuy committed:
Merge branch 'dev' into add-timeseries-to-datapackage

# Conflicts:
#	data_adapter_oemof/adapters.py

henhuy committed:
Rename test structures folder

henhuy committed:
Add structures folder to .gitignore

henhuy committed:
Remove test test

henhuy committed:
Add Adapters to BUS_NAME_MAP.yaml

henhuy committed:
Refactor default parameter function in Adapter class

henhuy committed:
Refactor build_datapackage to iterate over components

henhuy committed:
Minor change

henhuy committed:
Add timeseries to mapper; search timeseries for map keys

henhuy committed:
Change test example to hack-a-thon collection

henhuy commented:
Did some refactoring - but not finished yet (switched to draft mode). Here my ideas/changes:
- I refactored mapping to read TS related keys either from mapping or to guess it (if only one timeseries exists).
- Mapping has to be improved, so that process-depended mappings are allowed.
- Removed "profile", "region" and "year" mappings in `parametrize_dataclass` as "profile" will never be found in data and "region" and "year" are not in facades (must be handled differently)
- switched to hack-a-thon example, as minimal example is not like real SEDOS data (SEDOS data does not contain carrier, tech or profile - aka is much harder to parse)
- refactor_timeseries is now done every time - as timeindex is needed and regions have to be exploded always
- unfortunately `refactor_timeseries` currently fails for "onshore" of wind turbine process (I think the timeindex/region processing can be simplified - looks too complicated)
- hardcoded "carrier" and "tech" for facades - we can think about smart entries there later
- could not find package for `setup_environment`? (I thought I only commmented it out - but seems to be gone)

Hope you can understand my ideas (as I wrote them down very quickly - sorry for that!)

henhuy commented:
Also as a recommendation:
- Write smaller tests!  `test_example` with `build_datapackage` is way too heavy to test :) (for example see `test_refactor_timeseries`)
- use pre-commit for linting

henhuy committed:
Add test for timeseries refactoring

henhuy committed:
Simplify test for timeseries refactoring

henhuy committed:
Simplify timeseries refactoring

henhuy commented:
Started to refactor timeseries refactor function. Not yet ready.
Following things must be done:
Pivot region to every column
explode columns
append dataframes with different timeindex using axis=0

If the tests runs successfully, this should be solved!

FelixMau committed:
Pivot region to every column
explode columns
append dataframes with different timeindex using axis=0

FelixMau commented:
Thank you for your review and sorry for untidy code! My pre-commit checks are not automatically running, sometimes I forget to run them.

I hope I could simplify the `refactor_timeseries` process, yet It is somehow complicated to catch and map everything. 
Maybe I have overseen the correct Pandas functionality, I agree there must be some better way to directly map a new index and Pivot directly from `explode` but I could not ind it. I think using `MultiIndexing` together with `stacking`/`unstacking` would work as well but I think this approach is more complicated to understand.

Regarding `setup_environment` please see #19 

Can you please explain further how mapping should be improved? Do you want the mapper to be able to handle the same facade differently for different Processes?

FelixMau committed:
Pre-commit

FelixMau committed:
Change order (test failed)

FelixMau committed:
Change order (test failed)

FelixMau committed:
Adding defaults for non kwarg arguments. Class creation would fail otherwise.
Note: Experimental changes only. Default entries must be found within data or come from another source since those entries are fundamental for facades to work propperly.

Commit:
Sort columns before assertion

This is to prevent assertion error just because of
wrong column order.

FelixMau committed:
Adding defaults for non kwarg arguments. Class creation would fail otherwise.
Note: Experimental changes only. Default entries must be found within data or come from another source since those entries are fundamental for facades to work propperly.

Commit:
Add further comments to refactor_timeseries

Commit:
Refactor concatination

henhuy commented on data_adapter_oemof/adapters.py:
```diff
@@ -313,7 +328,21 @@ def parametrize_dataclass(
+                         "condensing_efficiency": 1,
+                         "electric_efficiency": 1,
+                         "thermal_efficiency": 1,
+                         }:
```
I disagree with adding all those default kwargs.
The whole point of the Adapter class was to have as minimal effort to add new Adapters.
But instead to compare give data with related Facade and try to match it automatically, maybe make some smart guesses about the busses.
Having this implementation, we have lots of extra work (we don't have time for)!

henhuy reviewed (changes requested):
Also many of this changes are unrelated to timeseries issue.
Please let us move to smaller but clearer PRs.

henhuy committed:
Set default mappings for carrier and tech

henhuy committed:
Add type to adapters and remove unnecessary code

henhuy committed:
Merge branch 'dev' into add-timeseries-to-datapackage

# Conflicts:
#	data_adapter_oemof/adapters.py
#	data_adapter_oemof/build_datapackage.py
#	data_adapter_oemof/mappings.py
#	data_adapter_oemof/mappings/BUS_NAME_MAP.yaml
#	tests/test_example.py

henhuy committed:
Minor change

FelixMau committed:
Adding test for get_sequence mapper

FelixMau committed:
Test getting sequence names
